fix(build): building on Cygwin with bundled dependencies

ptomulik opened the pull request:
Fixes compilation/link/installation on Cygwin with bundled dependencies (cmake.deps). Only builds with LuaJIT are fixed. The main intent was to allow Cygwin users to just "make && make install" (with all defaults). Compilation with PUC LUA appeared to be much harder to fix, so perhaps separate Issue/PR for this shall be an option.

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

ptomulik committed:
fix(build): build on Cygwin (with bundled luajit)

Fixes compilation/link/installation on Cygwin with bundled dependencies
(cmake.deps). Only builds with LuaJIT are fixed. Linking against PUC LUA still
does not work and seems to be much harder to fix.

dundargoc commented:
I see in https://github.com/LuaJIT/LuaJIT/commit/55a42da36ea6dd357ef928dd5c6776278b32ab92 that luajit has dropped support for cygwin, which makes me hesitant to include this. We could merge this on a "if it works, it works" basis but I suspect it will stop working in the near future anyway. I'd lean towards not merging this, but don't feel strongly enough to block if @justinmk or @jamessan like this.

dundargoc commented on cmake.deps/cmake/BuildLuajit.cmake:
```diff
@@ -41,7 +41,16 @@ if(APPLE)
+    COMMAND ${CMAKE_COMMAND} -E copy ${DEPS_BUILD_DIR}/src/luajit/src/libluajit-5.1.dll.a ${DEPS_LIB_DIR}
+    # Luarocks searches for lua51.dll in lib
+    COMMAND ${CMAKE_COMMAND} -E copy ${DEPS_BUILD_DIR}/src/luajit/src/cyglua51.dll ${DEPS_LIB_DIR})
+
```
I think this is missing the `CopyFilesGlob.cmake` commands the other windows builds have for the extra `share` files luajit ships with. I don't think you'll notice them missing unless you specifically try to call on a function from them like in `p.lua`.

dundargoc commented on src/nvim/CMakeLists.txt:
```diff
@@ -816,6 +816,14 @@ if(EXISTS ${DEPS_PREFIX}/lib/nvim/parser)
     COMMAND ${CMAKE_COMMAND} -E ${COPY_DIRECTORY} ${DEPS_PREFIX}/lib/nvim/parser ${BINARY_LIB_DIR}/parser)
 endif()
 
+if(CYGWIN AND (EXISTS ${DEPS_PREFIX}/bin/cyglua51.dll))
```
Kinda hacky, I think the "correct" way to do this would be to search for `cyglua51.dll` in `cmake/FindLuajit.cmake` and drop something like `list(APPEND LUAJIT_NAMES cyglua51)` (that or maybe it's `libcyglua511`).

dundargoc commented:
Wait, hold up. I just reread the commit message. What do you mean make it buildable without depending on `cmake.deps`? You're modidying a file in `cmake.deps` though?

justinmk commented:
> I see in [LuaJIT/LuaJIT@55a42da](https://github.com/LuaJIT/LuaJIT/commit/55a42da36ea6dd357ef928dd5c6776278b32ab92) that luajit has dropped support for cygwin,

Thanks for mentioning that.
Does building Nvim with PUC Lua on cygwin work, currently? We should focus on that. And show an error or at least a warning if trying to build with cygwin + luajit. And update our docs with guidance.

dundargoc commented:
> Does building Nvim with PUC Lua on cygwin work, currently?

@ptomulik can you confirm? Build instructions on how to use Lua instead of Luajit should be in `build.md`.

ptomulik commented:
> Wait, hold up. I just reread the commit message. What do you mean make it buildable without depending on `cmake.deps`?

Current docs (BUILD.md) state that Cygwin build requires user to install manually build/runtime dependencies, for example from Cygwin packages repo (setup.exe), and then compile neovim. But most of these deps can be downloaded and built locally by CMake scripts located under cmake.deps. That what I meant saying "without the need of manually installing the dependencies defined by cmake.deps" - there is no need to go though the manual installation via setup.exe for packages that are handled by cmake.deps. The method of "make && make install" shall just work now. 

>  You're modidying a file in `cmake.deps` though?

Only BuildLuajit.cmake.

BTW, some build/runtime deps are now missing from Cygwin repository and can no longer be conveniently installed via setup.exe.

justinmk commented:
Waiting for @dundargoc 's comments to be addressed. Also please explain a bit more what was done here, not just "Makes neovim buildable".

ptomulik commented:
> > Does building Nvim with PUC Lua on cygwin work, currently?
> 
> @ptomulik can you confirm? Build instructions on how to use Lua instead of Luajit should be in `build.md`.

Long story short: PUC LUA doesn't work for me, and I wasn't able fix this now. To summarize the problems I found:

1. Unresolved externals (``lua_*`` symbols) when linking ``cygnlua0.dll``.
2. Added ``liblua.a`` to list of libraries when linking ``cygnlua0.dll`` => links correctly, but yields ``error loading module nlua0 [...] dynamic libraries not enabled``.
3. Added ``-DLUA_USE_DLOPEN`` to ``LUA_CFLAGS`` => ``dynamic libraries not enabled`` gone, but now I have either SEGFAULTs or ABRT.

I think it's best to leave building against PUC LUA for separate issue/PR.

justinmk commented:
> I think it's best to leave building against PUC LUA for separate issue/PR.

This is strange/suspicious. PUC Lua should be easier to build, not harder. Are you using 5.1  (not a newer version) ?

Meanwhile, I am going to YOLO this PR since all the related bits are guarded by `if(CYGWIN)`.
